[Mobile Payments] [SDK 2] Switch to SDK 2

allendav opened the pull request:
This PR merges the long-lived feature/stripe-sdk-2 branch to develop. This will transition the app from SDK 1.x to SDK 2.x. This adds support for the M2 reader, changes the reader software updates to include mandatory un-cancellable updates, and other fixes.

To Test:

- Ensure you can
    - connect to a Chipper reader
    - connect to a M2 reader - see known issue #5326
    - choose from multiple discovered readers
    - connect to a reader after changing your store’s address in wp-admin (e.g. street address)
    - update a reader’s software - see known issues #5324 and #5325 
    - gracefully handle a reader with too low a battery to update - see known issue #5327
    - collect a tap payment
    - collect a chip (insert) payment
    - collect a swipe payment

Update release notes:
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ctarda committed:
Bump Stripe Terminal SDK to 2.0

ctarda committed:
Solve the easiest compilation errors, renaming classes

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

ctarda committed:
Update error enumeration

ctarda committed:
Update conformance to discovery and software updates

ctarda committed:
Expose location ID

ctarda committed:
Pass location id when connecting to a reader

ctarda committed:
Comment out software updates. We will need to come back to this later

ctarda committed:
Comment logic to trigger updates

ctarda committed:
Update signature of collectPaymentMethod

ctarda committed:
Bring logic to install an optional update back

ctarda committed:
Build support for software updates within the existing APIs

ctarda committed:
And make tests pass

ctarda committed:
Add basic logic to fetch a reader location

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

# Conflicts:
#	Podfile.lock

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2. Add support for M2 reader

# Conflicts:
#	Podfile.lock

ctarda committed:
Add support for M2 reader

ctarda committed:
Add test for M2 reader type

ctarda committed:
Fix test

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

# Conflicts:
#	Podfile
#	Podfile.lock

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

# Conflicts:
#	Podfile.lock

ctarda committed:
Capture a location id from a mock service, if it is the first connection to a givven reader

ctarda committed:
Update location id

ctarda committed:
Add traces

ctarda committed:
Cleanup

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

ctarda committed:
Extract obtention of configuration to a separate method

ctarda committed:
Attempt to map obtation of configuration to connection

ctarda committed:
Compose two Futures. First, obtain location id then connect to reader

ctarda committed:
Merge branch 'develop' into feature/stripe-sdk-2

ctarda committed:
Update signature of test mock

koke committed:
First iteration on new UI for mandatory updates

koke committed:
Allow required updates to be canceled

koke committed:
Animate illustration on installation progress

koke committed:
Merge branch 'develop' into feature/stripe-sdk-2

koke committed:
Merge branch 'feature/stripe-sdk-2' into feature/stripe-sdk-2-updates

koke committed:
Fix unit tests

koke committed:
Merge branch 'feature/stripe-sdk-2' into feature/stripe-sdk-2-updates

koke committed:
Fix bad merge

koke committed:
Rename cancelable type to avoid confusion

koke committed:
Improve documentation

koke committed:
More missing documentation

koke committed:
Rename cancelable alias to match the type on Hardware

koke committed:
Introduce some constants for clarity in UpdateProgressImage

koke committed:
Make optional updates work with the new UI

koke committed:
Ensure animation progress is between 0 and 1

koke committed:
Use ternary for actionsMode

koke committed:
Add missing image tests

koke committed:
Observe software update state through stores and actions

koke committed:
Use semantic colors for progress fill

koke committed:
Make arrow properly centered

koke committed:
Don't show message when update is complete

ctarda committed:
Bump Stripe Terminal to 2.2

koke committed:
Merge pull request #4974 from woocommerce/feature/stripe-sdk-2-updates

[Mobile Payments] Firmware updates for StripeTerminal 2.0

koke committed:
Merge branch 'feature/stripe-sdk-2' into feature/stripe-sdk-2-updates-optional

koke committed:
Remove use of UpdateViewController

koke committed:
Show different messaging for required and optional updates

ctarda committed:
Merge pull request #5038 from woocommerce/feature/stripe-sdk-2-2-0

[Mobile Payments] Bump Stripe Terminal to 2.2

koke committed:
Remove all usage of checkForCardReaderUpdate

koke committed:
Merge branch 'feature/stripe-sdk-2' into feature/stripe-sdk-2-updates-optional

koke committed:
Simplify installUpdate API since all the callbacks come from softwareUpdateEvents now

koke committed:
Skip failure handling on update tests until it's implemented

koke committed:
Merge pull request #5027 from woocommerce/feature/stripe-sdk-2-updates-optional

[Mobile Payments] Make optional updates work with the new UI

koke committed:
Remove updates from card reader events

koke committed:
Include failure case in state for software update

koke committed:
Revert "Include failure case in state for software update"

This reverts commit 95d9fdb57ee1dd06fe5f7fdaa531dc2d56c71f97.

koke committed:
Handle update errors due to low battery

koke committed:
Add tests for update failure on low battery

allendav committed:
Add fetching location

koke committed:
Merge pull request #5073 from woocommerce/feature/stripe-sdk-2-error-handling

[Mobile Payments] Show error when mandatory update fails because of low battery

koke committed:
Handle errors different than low battery on required updates

koke committed:
Handle update errors in optional updates

koke committed:
Initialize modals so that they never show the default text from Storyboards

koke committed:
Add custom modal for non-retryable update errors

allendav committed:
Avoid using cached location

koke committed:
Add missing mock method

koke committed:
Merge pull request #5109 from woocommerce/feature/stripe-sdk-2-error-handling

[Mobile Payments] Handle software update errors for all cases

koke committed:
Revert "Add activity indicator view"

This reverts commit d0dc00b29df1e4fdeb1fe27a83159142a8dc6660.

koke committed:
Revert "Disable disconnect button during updates"

This reverts commit 205f29ac124f32d6d28cea2643c747e53e4d756c.

koke committed:
Merge branch 'develop-merge-sdk-2' into feature/stripe-sdk-2

koke committed:
Don't flag canceled updates as failed

koke committed:
Clear update errors after closing the modal

allendav committed:
Combine return and promise in two places

allendav committed:
Merge pull request #5087 from woocommerce/issue/5078-fetch-location

[Mobile Payments] [SDK 2] Add fetching actual location

koke committed:
Merge pull request #5126 from woocommerce/issue/5124-optional-update-cancel

[Mobile Payments] [SDK 2.0] Prevent errors when canceling updates

koke committed:
Merge branch 'feature/stripe-sdk-2' into feature/stripe-sdk-2-develop-merge

koke committed:
Merge branch 'develop' into feature/stripe-sdk-2-develop-merge

koke committed:
Fix unit tests after merge

allendav committed:
Bump minimum WCPay version, update merchant messaging

allendav committed:
Update plugin versions in unit tests

allendav committed:
Merge pull request #5166 from woocommerce/issue/5077-3-1-0

[Mobile Payments] [SDK 2] Bump minimum WCPay version, update merchant messaging

allendav committed:
Remove string args that are never used by view controller during collectPayment

allendav committed:
Remove onClearMessage logic, as it was not really connected to whether a chip card
was inserted; add in delegate for didReportReaderEvent that does report chip card
insertion status

allendav committed:
PaymentCaptureOrchestrator doesn't actually use displayMessage to display a message
from the card reader, so remove that for now for clarity's sake

allendav committed:
Add card removal to card reader service capture payment promise chain

allendav committed:
Get rid of obsolete remove card modal, add unit test for display message modal

allendav committed:
Remove prints

allendav committed:
Remove a few more prints :)

joshheald committed:
Include `onPresentMessage` argument label in call

Suggested code review change rather than using a comment to identify the first closure passed to `collectPayment`.

joshheald committed:
Fix whitespace

allendav committed:
Remove blank line after opening brace

allendav committed:
Rename onPresentMessage to onWaitingForInput to reduce confusion

allendav committed:
Replace Foundation Timer with Timer Publisher

allendav committed:
Merge pull request #5192 from woocommerce/issue/4430-remove-card

[Mobile Payments] [SDK 2] Ensure "Remove Card" prompt displays when card is left inserted in reader

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Fix compilation errors introduced by the merge conflict resolution process related to reader updates

allendav committed:
Revert low batt changes merged from develop

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Update logic for prompt and button for SDK2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge pull request #5240 from woocommerce/issue/5238-settings-update

[MobiIe Payments] [SDK 2] Sync prompt and update button in connected reader settings

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Bump to Stripe Terminal SDK version 2.3

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge pull request #5277 from woocommerce/issue/5142-sdk-2-3

[Mobile Payments] Bump to Stripe Terminal SDK version 2.3

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

allendav committed:
Merge branch 'develop' into feature/stripe-sdk-2

joshheald commented on Hardware/Hardware/CardReader/CardReaderConfigProvider.swift:
```diff
@@ -1,6 +1,16 @@
+    func fetchDefaultLocationID(completion: @escaping(String?, Error?) -> Void)
+}
+
+public protocol CardReaderConfigProvider: ReaderLocationProvider {
```
```suggestion
public protocol CardReaderConfigProvider: ReaderLocationProvider, ReaderTokenProvider {
```

🔧 It looks like this protocol also includes `ReaderTokenProvider`, but there's no explicit conformance to that one, I think we should add it.

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5424

joshheald commented on Hardware/Hardware/CardReader/CardReaderConfigProvider.swift:
```diff
@@ -1,6 +1,16 @@
+
+public protocol CardReaderConfigProvider: ReaderLocationProvider {
+    func fetchToken(completion: @escaping(String?, Error?) -> Void)
+    func fetchDefaultLocationID(completion: @escaping(String?, Error?) -> Void)
```
❓♻️   What do you think about using `Result<String, Error>` in our adapter? I can see the completion blocks are modelled on `ConnectionTokenCompletionBlock` from `StripeTerminal`, but we could potentially adapt those as well? 

Perhaps something to ticket for later...

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5425

joshheald commented on Hardware/Hardware/CardReader/CardReaderService.swift:
```diff
@@ -52,13 +52,8 @@ public protocol CardReaderService {
-    /// and it will complete when it's finished, unless there is any error.
-    func installUpdate() -> AnyPublisher<Float, Error>
+    /// To check the progress of the update, observe the softwareUpdateEvents publisher.
+    func installUpdate() -> Void
```
This is a nicer API 👍

joshheald commented on WooCommerce/Classes/ViewModels/CardPresentPayments/PaymentCaptureOrchestrator.swift:
```diff
@@ -207,7 +205,7 @@ private extension PaymentCaptureOrchestrator {
             switch result {
             case .success:
-                self?.celebrate()
+                self?.celebrate() // plays a sound, haptic
```
💭 Adding `let self = self` to the guard above would _probably_ be worthwhile here. I'm not sure we want to be doing the other things in this block but not saving the receipt, if the `PaymentCaptureOrchestrator` has gone away anyway... 

I traced that back as best I could see and while the `onCompletion` didn't seem like it would currently do anything too surprising in that case, this does seem like somewhere that strange bugs could creep in.

What do you think? Playing devil's advocate I can see the merits of trying to handle the success even if we can't save the receipt too... but generally I'd fall on the side of failing fast(er.)

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5426

joshheald commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsConnectedViewController.swift:
```diff
@@ -66,20 +58,19 @@ final class CardReaderSettingsConnectedViewController: UIViewController, CardRea
     }
 
+    func isReaderUpdateAvailable() -> Bool {
+        guard let viewModel = viewModel else {
+            return false
+        }
+        return viewModel.readerUpdateAvailable == true
```
```suggestion
        return viewModel?.readerUpdateAvailable == true
```

This could also be inlined at the call sites of this function, if desired...

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5427

joshheald commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/CardReadersV2/CardReaderSettingsConnectedViewModel.swift:
```diff
@@ -78,64 +114,32 @@ final class CardReaderSettingsConnectedViewModel: CardReaderSettingsPresentedVie
+    func cancelCardReaderUpdate() {
+        softwareUpdateCancelable?.cancel(completion: { [weak self] result in
+            if case .failure(let error) = result {
+                print("=== error canceling software update: \(error)")
```
```suggestion
                DDLogError("=== error canceling software update: \(error)")
```

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5428

joshheald reviewed (approved):
Some minor suggestions added, all optional.

Testing progress
- [x] connect to a Chipper reader
- [ ] ~connect to a M2 reader~
- [ ] ~choose from multiple discovered readers~
- [x] connect to a reader after changing your store’s address in wp-admin (e.g. street address)
- [ ] update a reader’s software - done on sim, todo on physical reader after 👇 
- [ ] gracefully handle a reader with too low a battery to update - 🕐 trying to drain the battery to test this! 🔋 
- [x] collect a tap payment
- [x] collect a chip (insert) payment
- [ ] ~collect a swipe payment~

Strikethrough denotes tests I can't do for lack of equipment. @ctarda are there any of those you can do?

joshheald commented:
@allendav I found #5332 during testing. It's a bit of an edge case, so we could consider merging even with this issue. It seems like a completion block is being called when it shouldn't be, and also might be solved with a `guard self = self` in the right block, as I suspect the buttons don't work because of referencing something else on `self`. That's all guesswork at the moment I'm afraid though!

allendav commented:
Thank you for the review @joshheald ! In the interest of avoiding merge conflicts, would you be ok with me addressing the feedback in follow on issues? I'd like to merge this as quickly as possible.

joshheald commented:
> Thank you for the review @joshheald ! In the interest of avoiding merge conflicts, would you be ok with me addressing the feedback in follow on issues? I'd like to merge this as quickly as possible.

@allendav Yeah no problem at all... I should have called more attention to it but I meant that as one of your options when I wrote "Some minor suggestions added, *all optional*."
